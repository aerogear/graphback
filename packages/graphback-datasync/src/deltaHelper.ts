import { ModelDefinition, getTableName } from '@graphback/core';
import { Db, Collection, ObjectId } from "mongodb";
import { DataSyncFieldNames } from './util';

export function getDeltaTableName(tableName: string) {
  return `${tableName}Delta`
}

/**
 * M
 */
export class MongoDeltaHelper {
  protected db: Db;
  protected collectionName: string;
  public constructor(model: ModelDefinition, db: Db) {
    this.db = db;
    this.collectionName = getDeltaTableName(getTableName(model.graphqlType));
  }

  public async insertDiff(updatedDocument: any) {
    const version = updatedDocument[DataSyncFieldNames.version];
    const { _id } = updatedDocument;

    const diff: any = {
      docId: _id,
      [DataSyncFieldNames.version]: version,
      document: updatedDocument
    }
    await this.db.collection(this.collectionName).insertOne(diff);
  }

  public async findBaseForConflicts(updateDocument: any) {
    if (updateDocument._id === undefined || updateDocument[DataSyncFieldNames.version] === undefined) {
      throw new Error('Both _id and version field are needed for finding base');
    }

    const filter = {
      docId: updateDocument._id,
      [DataSyncFieldNames.version]: updateDocument[DataSyncFieldNames.version]
    };

    const result = await this.db.collection(this.collectionName).findOne(filter);

    return result?.document;
  }
}