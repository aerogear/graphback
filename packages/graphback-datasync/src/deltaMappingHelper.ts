import * as pluralize from "pluralize";
export function getDeltaListType(typeName: string): string {
  return `${typeName}DeltaList`;
}
export function getDeltaType(typeName: string) {
  return `${typeName}Delta`;
}
/**
 * Get the name of delta query for a model
 * @param typeName Name of the model
 */
export function getDeltaQuery(typeName: string) {
  return `sync${pluralize(typeName)}`;
}