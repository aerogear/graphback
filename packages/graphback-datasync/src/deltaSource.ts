import { ModelDefinition, getTableName } from '@graphback/core';
import { Db } from "mongodb";
import { applyIndexes } from '@graphback/runtime-mongo';
import { DataSyncFieldNames } from './util';

export function getDeltaTableName(tableName: string) {
  return `${tableName}Delta`
}

/**
 * Provides the ability to insert delta snapshots into MongoDB collections
 * and get base for conflict resolution
 */
export class MongoDeltaSource<Type = any> {
  protected db: Db;
  protected collectionName: string;
  protected deltaTTL: number;
  public constructor(model: ModelDefinition, db: Db, deltaTTL: number) {
    this.db = db;
    this.collectionName = getDeltaTableName(getTableName(model.graphqlType));
    this.deltaTTL = deltaTTL;
    applyIndexes([
      {
        key: {
          [DataSyncFieldNames.ttl]: 1
        },
        expireAfterSeconds: 0
      }
    ], this.db.collection(this.collectionName)).catch((e: any) => {
      // eslint-disable-next-line no-console
      console.error(`Could not create TTL Index for delta table: ${this.collectionName}: ${e}`)
    })
  }

  public async insertDiff(updatedDocument: Type) {
    const { _id, [DataSyncFieldNames.version]: version } = (updatedDocument as any);

    const diff: any = {
      docId: _id,
      [DataSyncFieldNames.version]: version,
      document: updatedDocument,
      [DataSyncFieldNames.ttl]: new Date(Date.now() + (this.deltaTTL * 1000))
    }
    await this.db.collection(this.collectionName).insertOne(diff);
  }

  public async findBaseForConflicts(updateDocument: any): Promise<Type> {
    if (!updateDocument._id || !updateDocument[DataSyncFieldNames.version]) {
      throw new Error(`Both _id and ${DataSyncFieldNames.version} field are needed for finding base`);
    }

    const filter = {
      docId: updateDocument._id,
      [DataSyncFieldNames.version]: updateDocument[DataSyncFieldNames.version]
    };

    const result = await this.db.collection(this.collectionName).findOne(filter);

    return result?.document;
  }
}
