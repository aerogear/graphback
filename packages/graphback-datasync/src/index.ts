export * from './DataSyncPlugin'
export * from './providers';
export * from './services';
export * from './util';
export * from './helpers/createDataSync';
export * from './deltaSource';

//Required for plugins
export { DataSyncPlugin as Plugin } from './DataSyncPlugin'