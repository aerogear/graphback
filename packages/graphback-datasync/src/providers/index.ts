export * from "./DataSyncProvider";
export * from './DatasyncMongoDBDataProvider';