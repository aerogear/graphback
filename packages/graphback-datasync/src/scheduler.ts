import * as Agenda from "agenda";
import { Db, MongoClient } from "mongodb";

export interface PruneConfiguration {
  agendaConfig?: Agenda.AgendaConfiguration
  pruneInterval?: string | number
}

export interface TTLMap {
  [modelName: string]: number
}

/**
 * A class for creating prune jobs
 */
export class PruneScheduler {
  public objectDb: Db;
  public pruneInterval: string | number;
  public agenda: Agenda;

  public constructor(config: PruneConfiguration, db: Db) {
    this.agenda = new Agenda(config.agendaConfig);
    if (config.agendaConfig?.db === undefined && config.agendaConfig?.mongo === undefined) {
      this.agenda.mongo(db);
    }
    this.agenda.start().then(() => {

      // Cleanup ensure job stopped running before exiting
      const graceful = async () => {
        await this.agenda.stop();
        process.exit(0);
      }

      process.on('SIGTERM', graceful);
      process.on('SIGINT' , graceful);
    }).catch((e: any) => {
      console.error('Error starting scheduler: ', e);
    })
    this.objectDb = db;
    this.pruneInterval = config.pruneInterval || '1 minute'
  }

  public defineAndStartPruneJob(collectionName: string, TTL: number) {
    const pruneJobName = this.definePruneJob(collectionName, TTL);
    this.startPruneJob(pruneJobName);
  }

  public definePruneJob(collectionName: string, TTL: number): string {
    const pruneJobName = `prune:${collectionName}`;
    const db = this.objectDb
    this.agenda.define(pruneJobName, async (_:any, done: (e: any) => void) => {
      db.collection(collectionName).deleteMany({
        _deleted: true,
        updatedAt: {
          $lte: Date.now() - (TTL * 1000) // TTL is in seconds
        }
      })
        .then(() => {
          done(undefined);
        })
        .catch((e: any) => {
          console.error('Error: Prune job failed: ', e);
          done(e);
        });

    });
  
    return pruneJobName
  }

  public startPruneJob(jobName: string) {
    this.agenda.every(this.pruneInterval, jobName).catch((e: any) => {
      console.error(`Error: Scheduling ${jobName} failed: ${e}`);
    });
  }
}