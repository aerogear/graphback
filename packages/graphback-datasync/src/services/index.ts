export * from './createDataSyncServices';
export * from './DataSyncCRUDService'