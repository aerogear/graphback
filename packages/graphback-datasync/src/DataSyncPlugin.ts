import { GraphQLNonNull, GraphQLSchema, buildSchema, GraphQLResolveInfo, GraphQLInt, GraphQLBoolean, GraphQLList, GraphQLObjectType, GraphQLField } from 'graphql';
import { GraphbackCoreMetadata, GraphbackPlugin, ModelDefinition, getInputTypeName, GraphbackOperationType, parseRelationshipAnnotation, GraphbackContext, getSelectedFieldsFromResolverInfo, GraphbackTimestamp, isSpecifiedGraphbackScalarType } from '@graphback/core'
import { SchemaComposer } from 'graphql-compose';
import { IResolvers, IFieldResolver } from '@graphql-tools/utils'

import { getDeltaType, getDeltaListType, getDeltaQuery } from "./deltaMappingHelper";
import { isDataSyncService, isDataSyncModel, DataSyncFieldNames, GlobalConflictConfig, getModelConfigFromGlobal } from "./util";

export const DATASYNC_PLUGIN_NAME = "DataSyncPlugin";

export interface DataSyncPluginConfig {
  conflictConfig: GlobalConflictConfig
}

/**
 * DataSync plugin
 *
 * Plugin is enabled by """ @datasync """ annotation
 * It will generate delta queries
 */
export class DataSyncPlugin extends GraphbackPlugin {
  protected config: DataSyncPluginConfig;

  public constructor(config?: DataSyncPluginConfig) {
    super()
    this.config = {
      conflictConfig: {},
      ...config
    };
  }

  public transformSchema(metadata: GraphbackCoreMetadata): GraphQLSchema {
    const schema = metadata.getSchema()
    const schemaComposer = new SchemaComposer(schema);
    if (!schemaComposer.has(GraphbackTimestamp.name)) {
      schemaComposer.createScalarTC(GraphbackTimestamp);
    }
    const models = metadata.getModelDefinitions();
    if (models.length === 0) {
      this.logWarning("Provided schema has no models. Returning original schema without any changes.")

      return schema;
    }
    let dataSyncModelCount = 0;
    models.forEach((model: ModelDefinition) => {
      // Diff queries
      if (isDataSyncModel(model)) {

        dataSyncModelCount+= 1;

        this.addDataSyncFieldsToModel(schemaComposer, model);

        this.addDataSyncFieldsToInputTypes(schemaComposer, model);

        this.addDeltaQuery(schemaComposer, model);
      }
    })

    if (dataSyncModelCount === 0) {
      console.warn("No DataSync Models detected, ensure that your models are properly annotated.")
    }

    return buildSchema(schemaComposer.toSDL())
  }

  /**
   * Creates resolvers for Data Synchonization
   *
   * @param {GraphbackCoreMetadata} metadata - Core metatata containing all model information
   */
  public createResolvers(metadata: GraphbackCoreMetadata): IResolvers {
    const resolvers: IResolvers = {
      Query: {},
      Mutation: {},
      Subscription: {}
    }

    const models = metadata.getModelDefinitions()

    for (const model of models) {
      // If delta marker is encountered, add resolver for `delta` query
      if (isDataSyncModel(model)) {
        this.addDeltaSyncResolver(model, resolvers.Query as IFieldResolver<any, any>)
      }
    }

    return resolvers
  }

  public createResources(metadata: GraphbackCoreMetadata): void {
    // TODO generate delta resolvers
    // TODO DataSource support for deltas
    // Schema plugin is going to create schema for us
    // No work to be done
  }

  public getPluginName() {
    return DATASYNC_PLUGIN_NAME;
  }

  protected addDeltaSyncResolver(model: ModelDefinition, queryObj: IFieldResolver<any, any>) {
    const modelName = model.graphqlType.name;
    const deltaQuery = getDeltaQuery(modelName)

    queryObj[deltaQuery] = async (_: any, args: any, context: GraphbackContext, info: GraphQLResolveInfo) => {
      const dataSyncService = isDataSyncService(context.graphback.services[modelName]);
      if (dataSyncService === undefined) {
        throw Error("Service is not a DataSyncCRUDService. Please use DataSyncCRUDService and DataSync-compliant DataProvider with DataSync Plugin to get Delta Queries.")
      }

      const selectedFields = getSelectedFieldsFromResolverInfo(info, model, "items");
      const graphback = {
        services: context.graphback.services,
        options: { selectedFields }
      };

      return dataSyncService.sync(args.lastSync, {...context, graphback }, args.filter, args.limit);
    }
  }

  protected addDataSyncFieldsToModel(schemaComposer: SchemaComposer<any>, model: ModelDefinition) {
    const name = model.graphqlType.name;
    const modelTC = schemaComposer.getOTC(name);
    const TimestampSTC = schemaComposer.getSTC(GraphbackTimestamp.name);

    modelTC.addFields({
      [DataSyncFieldNames.lastUpdatedAt]: {
        type: TimestampSTC.getType(),
        description: "@index(name: 'Datasync_lastUpdatedAt')"
      }
    });

    const modelUsesVersion = getModelConfigFromGlobal(model.graphqlType.name, this.config.conflictConfig).enabled;
    if (modelUsesVersion) {
      modelTC.addFields({
        [DataSyncFieldNames.version]: {
          type: GraphQLInt
        }
      });
    }
  }

  protected addDataSyncFieldsToInputTypes(schemaComposer: SchemaComposer<any>, model: ModelDefinition) {

    // Add _version argument to UpdateInputType
    const updateInputType = schemaComposer.getITC(getInputTypeName(model.graphqlType.name, GraphbackOperationType.UPDATE));
    const modelUsesVersion = getModelConfigFromGlobal(model.graphqlType.name, this.config.conflictConfig).enabled;
    if (modelUsesVersion && updateInputType) {
      updateInputType.addFields({
        [DataSyncFieldNames.version]: {
          type: GraphQLNonNull(GraphQLInt)
        }
      });
    }
  }

  protected addDeltaQuery(schemaComposer: SchemaComposer<unknown>, model: ModelDefinition) {
    // Create Delta Type
    const modelName = model.graphqlType.name;
    const modelTC = schemaComposer.getOTC(modelName);
    const TimestampSTC = schemaComposer.getSTC(GraphbackTimestamp.name);

    const DeltaTypeFieldNames = this.getDeltaTypeFieldNames(modelTC.getType());

    // Add Delta Type to schema
    const DeltaOTC = modelTC.clone(getDeltaType(modelName));

    DeltaOTC.removeOtherFields(DeltaTypeFieldNames);

    DeltaOTC.setDescription(undefined);

    DeltaOTC.addFields({
      [DataSyncFieldNames.deleted]: GraphQLBoolean
    })

    // Create and Add Delta List type to schema
    const DeltaListOTC = schemaComposer.createObjectTC({
      name: getDeltaListType(modelName),
      fields: {
        items: GraphQLNonNull(GraphQLList(DeltaOTC.getType())),
        lastSync: GraphQLNonNull(TimestampSTC.getType()),
        limit: GraphQLInt
      }
    });

    // Add Delta Queries
    const deltaQuery = getDeltaQuery(model.graphqlType.name);
    schemaComposer.Query.addFields({
      [deltaQuery]: GraphQLNonNull(DeltaListOTC.getType())
    });

    const findFilterITC = schemaComposer.getITC(getInputTypeName(modelName, GraphbackOperationType.FIND));
    schemaComposer.Query.addFieldArgs(deltaQuery, {
      lastSync: GraphQLNonNull(TimestampSTC.getType()),
      filter: findFilterITC.getType(),
      limit: GraphQLInt
    });
  }

  private getDeltaTypeFieldNames(modelTC: GraphQLObjectType): string[] {
    const DeltaTypeFieldEntries = Object.entries(modelTC.getFields()).filter((e: [string, GraphQLField<any, any>]) => {
      // Remove relationship fields from delta Type
      return parseRelationshipAnnotation(e[1].description) === undefined;
    });

    return DeltaTypeFieldEntries.map(([fieldName, _]: [string, GraphQLField<any, any>]) => {
      return fieldName
    });
  }
}
