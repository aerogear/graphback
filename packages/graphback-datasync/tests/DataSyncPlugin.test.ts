import { readFileSync } from 'fs';
import { buildSchema, printSchema } from 'graphql';
import { GraphbackCoreMetadata, GraphbackPluginEngine } from '@graphback/core';
import { SchemaCRUDPlugin } from "@graphback/codegen-schema";
import { DataSyncPlugin } from '../src/DataSyncPlugin';

const schemaText = readFileSync(`${__dirname}/mock.graphql`, 'utf8')

test('Test snapshot config gql', async () => {
  const defautConfig = {
    "create": true,
    "update": true,
    "findOne": true,
    "find": true,
    "delete": true,
    "subCreate": true,
    "subUpdate": true,
    "subDelete": true
  }

  const schemaPlugin = new SchemaCRUDPlugin();
  const datasync = new DataSyncPlugin()
  const metadata = new GraphbackCoreMetadata({
    crudMethods: defautConfig
  }, buildSchema(schemaText))
  metadata.setSchema(schemaPlugin.transformSchema(metadata))
  const schema = datasync.transformSchema(metadata)
  expect(printSchema(schema)).toMatchSnapshot();
  const resolvers: any = datasync.createResolvers(metadata);
  expect(resolvers.Query.syncComments).toBeTruthy()
});

it('uses existing GraphbackTimestamp scalars', async () => {
  const defautConfig = {
    "create": true,
    "update": true,
    "findOne": true,
    "find": true,
    "delete": true,
    "subCreate": true,
    "subUpdate": true,
    "subDelete": true
  }

  const schemaPlugin = new SchemaCRUDPlugin();
  const datasync = new DataSyncPlugin()
  const metadata = new GraphbackCoreMetadata({
    crudMethods: defautConfig
  }, buildSchema(
    `
    """
    @model
    @datasync(
      ttl: 5184000
    )
    """
    type Comment {
      id: ID!
      title: String!
      description: String!
    }

    scalar GraphbackTimestamp
    `
  ))
  metadata.setSchema(schemaPlugin.transformSchema(metadata))
  const schema = datasync.transformSchema(metadata)
  expect(printSchema(schema)).toMatchSnapshot();
});

it('adds version when conflicts are enabled', async () => {
  const defautConfig = {
    "create": true,
    "update": true,
    "findOne": true,
    "find": true,
    "delete": true,
    "subCreate": true,
    "subUpdate": true,
    "subDelete": true
  }

  const schemaPlugin = new SchemaCRUDPlugin();
  const datasync = new DataSyncPlugin({ conflictConfig: { enabled: true } })
  const metadata = new GraphbackCoreMetadata({
    crudMethods: defautConfig
  }, buildSchema(
    `
    """
    @model
    @datasync(
      ttl: 5184000
    )
    """
    type Comment {
      id: ID!
      title: String!
      description: String!
    }

    scalar GraphbackTimestamp
    `
  ))
  metadata.setSchema(schemaPlugin.transformSchema(metadata))
  const schema = datasync.transformSchema(metadata)
  expect(printSchema(schema)).toMatchSnapshot();
});

test('When all CRUD flags are disabled, resolvers and root schema types are not generated', () => {
  const model = buildSchema(`
  """@model"""
  type Note {
    id: ID!
    title: String
  }`)

  const pluginEngine = new GraphbackPluginEngine({
    schema: model,
    plugins: [
      new DataSyncPlugin()
    ],
    config: {
      crudMethods: {
        find: false,
        findOne: false,
        create: false,
        update: false,
        delete: false,
        subCreate: false,
        subDelete: false,
        subUpdate: false
      }
    }
  });

  const metadata = pluginEngine.createResources();
  const schema = metadata.getSchema();
  const resolvers = metadata.getResolvers();

  expect(resolvers).toBeUndefined()
  expect(schema.getQueryType()).toBeUndefined();
  expect(schema.getMutationType()).toBeUndefined();
  expect(schema.getSubscriptionType()).toBeUndefined();
});
