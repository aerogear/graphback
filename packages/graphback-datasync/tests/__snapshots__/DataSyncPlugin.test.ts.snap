// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test snapshot config gql 1`] = `
"\\"\\"\\"
@model
@datasync
\\"\\"\\"
type Comment {
  id: ID!
  title: String!
  description: String!

  \\"\\"\\"@manyToOne(field: 'comments', key: 'noteId')\\"\\"\\"
  noteComment: Note!

  \\"\\"\\"@index(name: 'Datasync_lastUpdatedAt')\\"\\"\\"
  _lastUpdatedAt: GraphbackTimestamp
}

type CommentDelta {
  id: ID!
  title: String
  description: String
  noteId: ID

  \\"\\"\\"@index(name: 'Datasync_lastUpdatedAt')\\"\\"\\"
  _lastUpdatedAt: GraphbackTimestamp
  _deleted: Boolean
}

type CommentDeltaList {
  items: [CommentDelta]!
  lastSync: GraphbackTimestamp!
  limit: Int
}

input CommentFilter {
  id: IDInput
  title: StringInput
  description: StringInput
  noteId: IDInput
  and: [CommentFilter!]
  or: [CommentFilter!]
  not: CommentFilter
}

type CommentResultList {
  items: [Comment]!
  offset: Int
  limit: Int
  count: Int
}

input CommentSubscriptionFilter {
  and: [CommentSubscriptionFilter!]
  or: [CommentSubscriptionFilter!]
  not: CommentSubscriptionFilter
  id: IDInput
  title: StringInput
  description: StringInput
}

input CreateCommentInput {
  title: String!
  description: String!
  noteId: ID!
}

input CreateNoteInput {
  title: String!
  description: String!
  mainCommentId: ID!
}

\\"\\"\\"
The javascript \`Date\` as integer. Type represents date and time as number of milliseconds from start of UNIX epoch.
\\"\\"\\"
scalar GraphbackTimestamp

input IDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  in: [ID!]
}

input MutateCommentInput {
  id: ID!
  title: String
  description: String
  noteId: ID
}

input MutateNoteInput {
  id: ID!
  title: String
  description: String
  mainCommentId: ID
}

type Mutation {
  likeNote(id: ID!): Note!
  createNote(input: CreateNoteInput!): Note
  updateNote(input: MutateNoteInput!): Note
  deleteNote(input: MutateNoteInput!): Note
  createComment(input: CreateCommentInput!): Comment
  updateComment(input: MutateCommentInput!): Comment
  deleteComment(input: MutateCommentInput!): Comment
}

\\"\\"\\"
@model
@datasync
\\"\\"\\"
type Note {
  id: ID!
  title: String!
  description: String!

  \\"\\"\\"@oneToOne(field: 'mainComment', key: 'mainCommentId')\\"\\"\\"
  mainComment: Comment!

  \\"\\"\\"@oneToMany(field: 'noteComment', key: 'noteId')\\"\\"\\"
  comments(filter: CommentFilter): [Comment]!

  \\"\\"\\"@index(name: 'Datasync_lastUpdatedAt')\\"\\"\\"
  _lastUpdatedAt: GraphbackTimestamp
}

type NoteDelta {
  id: ID!
  title: String
  description: String
  mainCommentId: ID

  \\"\\"\\"@index(name: 'Datasync_lastUpdatedAt')\\"\\"\\"
  _lastUpdatedAt: GraphbackTimestamp
  _deleted: Boolean
}

type NoteDeltaList {
  items: [NoteDelta]!
  lastSync: GraphbackTimestamp!
  limit: Int
}

input NoteFilter {
  id: IDInput
  title: StringInput
  description: StringInput
  mainCommentId: IDInput
  and: [NoteFilter!]
  or: [NoteFilter!]
  not: NoteFilter
}

type NoteResultList {
  items: [Note]!
  offset: Int
  limit: Int
  count: Int
}

input NoteSubscriptionFilter {
  and: [NoteSubscriptionFilter!]
  or: [NoteSubscriptionFilter!]
  not: NoteSubscriptionFilter
  id: IDInput
  title: StringInput
  description: StringInput
}

input OrderByInput {
  field: String!
  order: SortDirectionEnum = ASC
}

input PageRequest {
  limit: Int
  offset: Int
}

type Query {
  getLikedNotes(id: ID!, names: [String]!): Note!
  getLikedComments(id: ID!, names: [String]!): Comment!
  getNote(id: ID!): Note
  findNotes(filter: NoteFilter, page: PageRequest, orderBy: OrderByInput): NoteResultList!
  getComment(id: ID!): Comment
  findComments(filter: CommentFilter, page: PageRequest, orderBy: OrderByInput): CommentResultList!
  syncNotes(lastSync: GraphbackTimestamp!, filter: NoteFilter, limit: Int): NoteDeltaList!
  syncComments(lastSync: GraphbackTimestamp!, filter: CommentFilter, limit: Int): CommentDeltaList!
}

enum SortDirectionEnum {
  DESC
  ASC
}

input StringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  in: [String!]
  contains: String
  startsWith: String
  endsWith: String
}

type Subscription {
  newNote(filter: NoteSubscriptionFilter): Note!
  updatedNote(filter: NoteSubscriptionFilter): Note!
  deletedNote(filter: NoteSubscriptionFilter): Note!
  newComment(filter: CommentSubscriptionFilter): Comment!
  updatedComment(filter: CommentSubscriptionFilter): Comment!
  deletedComment(filter: CommentSubscriptionFilter): Comment!
}
"
`;

exports[`adds version when conflicts are enabled 1`] = `
"\\"\\"\\"
@model
@datasync(
  ttl: 5184000
)
\\"\\"\\"
type Comment {
  id: ID!
  title: String!
  description: String!

  \\"\\"\\"@index(name: 'Datasync_lastUpdatedAt')\\"\\"\\"
  _lastUpdatedAt: GraphbackTimestamp
  _version: Int
}

type CommentDelta {
  id: ID!
  title: String
  description: String
  _version: Int!

  \\"\\"\\"@index(name: 'Datasync_lastUpdatedAt')\\"\\"\\"
  _lastUpdatedAt: GraphbackTimestamp
  _deleted: Boolean
}

type CommentDeltaList {
  items: [CommentDelta]!
  lastSync: GraphbackTimestamp!
  limit: Int
}

input CommentFilter {
  id: IDInput
  title: StringInput
  description: StringInput
  and: [CommentFilter!]
  or: [CommentFilter!]
  not: CommentFilter
}

type CommentResultList {
  items: [Comment]!
  offset: Int
  limit: Int
  count: Int
}

input CommentSubscriptionFilter {
  and: [CommentSubscriptionFilter!]
  or: [CommentSubscriptionFilter!]
  not: CommentSubscriptionFilter
  id: IDInput
  title: StringInput
  description: StringInput
}

input CreateCommentInput {
  title: String!
  description: String!
}

\\"\\"\\"
The javascript \`Date\` as integer. Type represents date and time as number of milliseconds from start of UNIX epoch.
\\"\\"\\"
scalar GraphbackTimestamp

input IDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  in: [ID!]
}

input MutateCommentInput {
  id: ID!
  title: String
  description: String
  _version: Int!
}

type Mutation {
  createComment(input: CreateCommentInput!): Comment
  updateComment(input: MutateCommentInput!): Comment
  deleteComment(input: MutateCommentInput!): Comment
}

input OrderByInput {
  field: String!
  order: SortDirectionEnum = ASC
}

input PageRequest {
  limit: Int
  offset: Int
}

type Query {
  getComment(id: ID!): Comment
  findComments(filter: CommentFilter, page: PageRequest, orderBy: OrderByInput): CommentResultList!
  syncComments(lastSync: GraphbackTimestamp!, filter: CommentFilter, limit: Int): CommentDeltaList!
}

enum SortDirectionEnum {
  DESC
  ASC
}

input StringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  in: [String!]
  contains: String
  startsWith: String
  endsWith: String
}

type Subscription {
  newComment(filter: CommentSubscriptionFilter): Comment!
  updatedComment(filter: CommentSubscriptionFilter): Comment!
  deletedComment(filter: CommentSubscriptionFilter): Comment!
}
"
`;

exports[`uses existing GraphbackTimestamp scalars 1`] = `
"\\"\\"\\"
@model
@datasync(
  ttl: 5184000
)
\\"\\"\\"
type Comment {
  id: ID!
  title: String!
  description: String!

  \\"\\"\\"@index(name: 'Datasync_lastUpdatedAt')\\"\\"\\"
  _lastUpdatedAt: GraphbackTimestamp
}

type CommentDelta {
  id: ID!
  title: String
  description: String

  \\"\\"\\"@index(name: 'Datasync_lastUpdatedAt')\\"\\"\\"
  _lastUpdatedAt: GraphbackTimestamp
  _deleted: Boolean
}

type CommentDeltaList {
  items: [CommentDelta]!
  lastSync: GraphbackTimestamp!
  limit: Int
}

input CommentFilter {
  id: IDInput
  title: StringInput
  description: StringInput
  and: [CommentFilter!]
  or: [CommentFilter!]
  not: CommentFilter
}

type CommentResultList {
  items: [Comment]!
  offset: Int
  limit: Int
  count: Int
}

input CommentSubscriptionFilter {
  and: [CommentSubscriptionFilter!]
  or: [CommentSubscriptionFilter!]
  not: CommentSubscriptionFilter
  id: IDInput
  title: StringInput
  description: StringInput
}

input CreateCommentInput {
  title: String!
  description: String!
}

\\"\\"\\"
The javascript \`Date\` as integer. Type represents date and time as number of milliseconds from start of UNIX epoch.
\\"\\"\\"
scalar GraphbackTimestamp

input IDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  in: [ID!]
}

input MutateCommentInput {
  id: ID!
  title: String
  description: String
}

type Mutation {
  createComment(input: CreateCommentInput!): Comment
  updateComment(input: MutateCommentInput!): Comment
  deleteComment(input: MutateCommentInput!): Comment
}

input OrderByInput {
  field: String!
  order: SortDirectionEnum = ASC
}

input PageRequest {
  limit: Int
  offset: Int
}

type Query {
  getComment(id: ID!): Comment
  findComments(filter: CommentFilter, page: PageRequest, orderBy: OrderByInput): CommentResultList!
  syncComments(lastSync: GraphbackTimestamp!, filter: CommentFilter, limit: Int): CommentDeltaList!
}

enum SortDirectionEnum {
  DESC
  ASC
}

input StringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  in: [String!]
  contains: String
  startsWith: String
  endsWith: String
}

type Subscription {
  newComment(filter: CommentSubscriptionFilter): Comment!
  updatedComment(filter: CommentSubscriptionFilter): Comment!
  deletedComment(filter: CommentSubscriptionFilter): Comment!
}
"
`;
